jax/experimental/shard_map.py

```diff
@@ -1300,6 +1300,7 @@ def _shard_map_partial_eval(trace, shard_map_p, f, tracers, mesh, in_names,
   in_pvals = [t.pval for t in tracers]
   in_knowns, in_avals, in_consts = pe.partition_pvals(in_pvals)
   unk_in_names, known_in_names = pe.partition_list(in_knowns, in_names)
+  used_names = {n for names in known_in_names for ns in names.values() for n in ns}
```

Comment:
Github reviews don't allow me to comment on lines outside of the patch, so I am leaving a comment here instead.
Given this change I think line 1505:
```
  residuals = [newvar(_unshard_aval(mesh, {0: (*mesh.axis_names,)}, var.aval))
               for var, w in zip(jaxpr_staged.invars[:num_res], which) if w]
```
should probably be updated to use `used_names` and not `mesh.axis_names`